### PR TITLE
Rework find projects endpoint

### PR DESCRIPTION
**Pull request by BelehovEgor**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/412

### CODE REVIEW

**Comment by BelehovEgor** [comment]
the test for FindCommand is very bad and strange, please check and suggest how to fix this gloom

File: src/ProjectService.Data/Interfaces/IProjectRepository.cs [hunk 1]
```diff
@@ -45,7 +46,7 @@ public interface IProjectRepository
         /// </summary>
         /// <param name="showNotActive">Do you want to show inactive projects?</param>
         /// <returns>All projects.</returns>
-        IEnumerable<DbProject> GetProjects(bool showNotActive);
+        //IEnumerable<DbProject> GetProjects(bool showNotActive);
```
Comment (TIove) [inline src/ProjectService.Data/Interfaces/IProjectRepository.cs:49]: why comment?

Comment (BelehovEgor) [reply src/ProjectService.Data/Interfaces/IProjectRepository.cs]: now it deleted, before i didn't want to do it

**Review by djino97 — APPROVED** [review]

File: src/ProjectService.Business/Commands/FindProjectsCommand.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        private readonly ILogger<FindProjectsCommand> _logger;
+        private readonly IProjectRepository _repository;
+        private readonly IFindProjectsResponseMapper _responseMapper;
+        private readonly IFindDbProjectFilterMapper _filterMapper;
+        private readonly IRequestClient<IFindDepartmentsRequest> _findDepartmentsRequestClient;
+        private readonly IRequestClient<IGetDepartmentsNamesRequest> _getDepartmentsRequestClient;
+
+        private IDictionary<Guid, string> FindDepartment(string departmentName, List<string> errors)
+        {
+            IDictionary<Guid, string> pairs = new Dictionary<Guid, string>();
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/FindProjectsCommand.cs:29]: renamed

File: src/ProjectService.Business/Commands/FindProjectsCommand.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        public ProjectsResponse Execute(FindProjectsFilter filter, int skipCount, int takeCount)
+        {
+            if (filter == null)
+            {
+                throw new ArgumentNullException(nameof(filter));
+            }
+
+            List<string> errors = new();
+
+            IDictionary<Guid, string> pairs = null;
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/FindProjectsCommand.cs:117]: renamed

File: src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs [hunk 4]
```diff
@@ -7,6 +7,6 @@ namespace LT.DigitalOffice.ProjectService.Mappers.ModelsMappers.Interfaces
     [AutoInject]
     public interface IProjectInfoMapper
     {
-        ProjectInfo Map(DbProject dbProject, string departmentName);
+        ProjectInfo Map(DbProject dbProject, string departmetnName);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs]: fixed

File: src/ProjectService.Models.Broker/Requests/IGetUserDataRequest.cs [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 namespace LT.DigitalOffice.Broker.Requests
 {
-    [AutoInjectRequest(nameof(RabbitMqConfig.GetUserDataEndpoint))]
+    [AutoInjectRequest(nameof(RabbitMqConfig.FindDepartmentsEndpoint))]
     public interface IGetUserDataRequest
```
Comment (borovHokage) [inline src/ProjectService.Models.Broker/Requests/IGetUserDataRequest.cs:8]: why you dont rename interface?

File: src/ProjectService.Business/Commands/FindProjectsCommand.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            {
+                _logger.LogError(exc, errorMessage);
+
+                errors.Add(errorMessage);
+            }
+
+            return pairs;
+        }
+
+        private IDictionary<Guid, string> GetDepartmentName(List<DbProject> dbProjects, List<string> errors)
```
Comment (KostinP) [inline src/ProjectService.Business/Commands/FindProjectsCommand.cs:61]: Rename to Departments

File: src/ProjectService.Data/ProjectRepository.cs [hunk 7]
```diff
@@ -14,6 +15,28 @@ public class ProjectRepository : IProjectRepository
...
+
+            if (!string.IsNullOrEmpty(filter.ShortName))
+            {
+                dbProjects = dbProjects.Where(u => u.ShortName.ToUpper().Contains(filter.ShortName.ToUpper()));
+            }
+
+            if (filter.IdNameDepartments != null)
+            {
+                dbProjects = dbProjects.Where(u => filter.IdNameDepartments.Keys.Contains(u.DepartmentId));
+            }
```
Comment (KostinP) [inline src/ProjectService.Data/ProjectRepository.cs:35]: Why not used LinqKit?

**Review by KostinP — APPROVED** [review]

File: src/ProjectService.Models.Broker/Requests/IGetUserDataRequest.cs (hunk 5 above)
Comment (BelehovEgor) [reply src/ProjectService.Models.Broker/Requests/IGetUserDataRequest.cs:8]: It is my mistake, I don't know why I changed it, I returned it as it was

File: src/ProjectService.Business/Commands/FindProjectsCommand.cs (hunk 6 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/FindProjectsCommand.cs:61]: renamed

File: src/ProjectService.Data/ProjectRepository.cs (hunk 7 above)
Comment (BelehovEgor) [reply src/ProjectService.Data/ProjectRepository.cs:35]: What do you want to see here?

Comment (KostinP) [reply src/ProjectService.Data/ProjectRepository.cs:35]: I'm not sure if the name and department will always be unique for the project. Why not search for both?

**Review by borovHokage — APPROVED** [review]

**Review by TIove — APPROVED** [review]

**Pull request merged**